window/src/os/wayland/window.rs

```diff
@@ -867,12 +868,31 @@ impl WindowOps for WaylandWindow {
         });
     }
 
-    fn get_clipboard(&self, _clipboard: Clipboard) -> Future<String> {
+    fn get_clipboard(&self, clipboard: Clipboard) -> Future<String> {
         let mut promise = Promise::new();
         let future = promise.get_future().unwrap();
         let promise = Arc::new(Mutex::new(promise));
         WaylandConnection::with_window_inner(self.0, move |inner| {
-            let read = inner.copy_and_paste.lock().unwrap().get_clipboard_data()?;
+            let conn = Connection::get().unwrap().wayland();
+            let selection = if let Clipboard::PrimarySelection = clipboard {
+                conn.primary_selection.as_ref()
+            } else {
+                None
+            };
+            let read = match selection {
+                None => inner.copy_and_paste.lock().unwrap().get_clipboard_data(),
+                Some(selection) => {
```

Comment:
is it possible to move some of the more verbose parts of the getting/setting code into the `CopyAndPaste` struct?
I feel like `PrimarySelection` could potentially merge with that struct and that `get_clipboard_data` could be passed the `Clipboard` type and internally do the right thing

Reply:
Sure, will do.  I'll see if I can find a bit of time for this over the next days 🙂

Reply:
@wez I've moved things around a bit.  There's still a bit of duplication because the clipboard and primary selection types are different, and it also looks as if the data device code for clipboard access directly uses wayland protocol types whereas the new primary selection code uses some higher-level interface from the wayland toolkit library, which uses a slight different flow to access the selection offer and to create the selection source.

I've pushed it as a separate commit to be able to go back to the earlier version if you have objections; if you agree with the latest change I can squash it all together and clean up.